File: zipper.py
class ListZipper:
    def __init__(self, left, focus, right):
        self.left = left
        self.focus = focus
        self.right = right

    def __str__(self):
        return f"Left: {self.left}, Focus: {self.focus}, Right: {self.right}"

    def move_left(self):
        if not self.left:
            return None  # Cannot move left if already at the beginning
        return ListZipper(self.left[:-1], self.left[-1], [self.focus] + self.right)

    def move_right(self):
        if not self.right:
            return None  # Cannot move right if already at the end
        return ListZipper(self.left + [self.focus], self.right[0], self.right[1:])

    def update_focus(self, value):
        return ListZipper(self.left, value, self.right)

    def to_list(self):
        return self.left + [self.focus] + self.right

# Example usage
lst = [1, 2, 3, 4]
zipper = ListZipper([], lst[0], lst[1:])
print(zipper)  # Left: [], Focus: 1, Right: [2, 3, 4]
zipper = zipper.move_right()
print(zipper)  # Left: [1], Focus: 2, Right: [3, 4]
zipper = zipper.update_focus(5)
print(zipper)  # Left: [1], Focus: 5, Right: [3, 4]
print(zipper.to_list())  # [1, 5, 3, 4]

File: Zipper.hs
import Criterion.Main
import GHC.Stats
import System.Mem

-- Binary tree data structure
data Tree a = Empty | Node a (Tree a) (Tree a) deriving (Show)

-- Breadcrumbs for navigation in the tree
data Crumb a = LeftCrumb a (Tree a) | RightCrumb a (Tree a) deriving (Show)
type Breadcrumbs a = [Crumb a]

-- Zipper type combining the current focus and breadcrumbs
type Zipper a = (Tree a, Breadcrumbs a)

-- Go to the left child of the current focus
goLeft :: Zipper a -> Maybe (Zipper a)
goLeft (Node x left right, bs) = Just (left, LeftCrumb x right : bs)
goLeft (Empty, _) = Nothing

-- Go to the right child of the current focus
goRight :: Zipper a -> Maybe (Zipper a)
goRight (Node x left right, bs) = Just (right, RightCrumb x left : bs)
goRight (Empty, _) = Nothing

-- Move back up to the parent node
goUp :: Zipper a -> Maybe (Zipper a)
goUp (_, []) = Nothing
goUp (t, LeftCrumb x right : bs) = Just (Node x t right, bs)
goUp (t, RightCrumb x left : bs) = Just (Node x left t, bs)

-- Modify the value at the current focus
modify :: (a -> a) -> Zipper a -> Zipper a
modify f (Node x left right, bs) = (Node (f x) left right, bs)
modify _ (Empty, bs) = (Empty, bs)

-- Attach a new subtree at the current focus
attach :: Tree a -> Zipper a -> Zipper a
attach t (_, bs) = (t, bs)

-- Start a zipper at the root of the tree
startZipper :: Tree a -> Zipper a
startZipper t = (t, [])


-- Define some sample trees to use for benchmarking
sampleTree :: Tree Int
sampleTree = Node 1 (Node 2 Empty Empty) (Node 3 Empty Empty)

-- Define some benchmark functions
benchmarkGoLeft :: Zipper Int -> Benchmark
benchmarkGoLeft z = bench "goLeft" $ whnf goLeft z

benchmarkGoRight :: Zipper Int -> Benchmark
benchmarkGoRight z = bench "goRight" $ whnf goRight z

benchmarkGoUp :: Zipper Int -> Benchmark
benchmarkGoUp z = bench "goUp" $ whnf goUp z

benchmarkModify :: Zipper Int -> Benchmark
benchmarkModify z = bench "modify" $ whnf (modify (+1)) z

printMemoryUsage :: IO ()
printMemoryUsage = do
  performGC  -- Trigger garbage collection to get accurate stats
  stats <- getRTSStats
  putStrLn $ "Memory in use: " ++ show (gcdetails_mem_in_use_bytes $ gc stats) ++ " bytes"


main :: IO ()
main = do
  -- Run benchmarks
  defaultMain
    [ bgroup "Zipper Operations"
        [ benchmarkGoLeft (startZipper sampleTree)
        , benchmarkGoRight (startZipper sampleTree)
        , benchmarkGoUp (startZipper sampleTree)
        , benchmarkModify (startZipper sampleTree)
        ]
    ]

  -- Print memory usage
  printMemoryUsage
